### PR TITLE
Kafka source Confluent schema registry connectivity and OAuth implementation

### PR DESCRIPTION
**Pull request by ajeeshakd**
### Description
- Implemented PLAIN and OAUTHBEARER authentication mechanisms
- Kafka source will be able to connect to the Local schema registry as well as the Confluent Cloud schema registry with authentication mechanisms.
- Implemented the Confluent cloud cluster connectivity
- Supported security protocols as follows
       PLAINTEXT - Un-authenticated, non-encrypted channel.
       SASL_PLAINTEXT - SASL authenticated, non-encrypted channel.
       SASL_SSL - SASL authenticated, SSL channel.

### Issues Resolved
- Github issue https://github.com/opensearch-project/data-prepper/issues/254
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java [hunk 1]
```diff
@@ -51,6 +51,26 @@ public class KafkaSourceConfig {
     @JsonProperty("acknowledgments_timeout")
     private Duration acknowledgementsTimeout = DEFAULT_ACKNOWLEDGEMENTS_TIMEOUT;
 
+    @JsonProperty("serde_format")
+    private String serdeFormat;
+
+    @JsonProperty("client_dns_lookup")
+    private String clientDnsLookup;
```
Comment (kkondaka) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java:58]: If we want to support this, we should make it enum that only accepts the values `default` or `use_all_dns_ips` or `resolve_canonical_bootstrap_servers_only`

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java [hunk 2]
```diff
@@ -51,6 +51,26 @@ public class KafkaSourceConfig {
...
+
+    @JsonProperty("ssl_endpoint_identification_algorithm")
+    private String sslEndpointIdentificationAlgorithm;
```
Comment (kkondaka) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java:61]: Why is this needed? Where is this used. I have added https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/InsecureSslEngineFactory.java that can be used for insecure client.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java [hunk 3]
```diff
@@ -268,47 +300,22 @@ private Properties getConsumerProperties(final TopicConfig topicConfig) {
             throw new RuntimeException("Bootstrap servers are not specified");
         }
         properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+            if (isKafkaClusterExists(sourceConfig.getBootStrapServers())) {
+                throw new RuntimeException("Can't be able to connect to the given Kafka brokers... ");
```
Comment (kkondaka) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java:304]: I don't think we should exit if we can't connect. We should retry.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java [hunk 4]
```diff
@@ -418,4 +425,168 @@ private static String readErrorMessage(InputStream errorStream) throws IOExcepti
...
+
+    private void isTopicExists(String topicName, String bootStrapServer, Properties properties) {
+        List<String> bootStrapServers = new ArrayList<>();
+        String servers[];
+        if (bootStrapServer.contains(",")) {
+            servers = bootStrapServer.split(",");
+            bootStrapServers.addAll(Arrays.asList(servers));
+        } else {
+            bootStrapServers.add(bootStrapServer);
+        }
```
Comment (kkondaka) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java:528]: I don't this is needed. You just pass the `bootStrapServers` as is with comma separated values in it. That should work.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/PlainTextAuthConfig.java [hunk 5]
```diff
@@ -19,6 +19,19 @@ public class PlainTextAuthConfig {
     @JsonProperty("password")
     private String password;
 
+    @JsonProperty("sasl_mechanism")
```
[the remark this answers is not in the document]
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/PlainTextAuthConfig.java]: Removed the sasl_mechanism config value and trying to get the same value from 
authentication:
   sasl:
      plaintext:

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java [hunk 6]
```diff
@@ -268,47 +300,22 @@ private Properties getConsumerProperties(final TopicConfig topicConfig) {
...
-            schemaType = getSchemaType(schemaConfig.getRegistryURL(), topicConfig.getName(), schemaConfig.getVersion());
-    }
-        if (schemaType.isEmpty()) {
-            schemaType = MessageFormat.PLAINTEXT.toString();
+        if (StringUtils.isNotEmpty(sourceConfig.getClientDnsLookup())) {
+            properties.put("client.dns.lookup", sourceConfig.getClientDnsLookup());
         }
-        setPropertiesForSchemaType(properties, schemaType);
+        if (StringUtils.isNotEmpty(sourceConfig.getSslEndpointIdentificationAlgorithm())) {
+            properties.put("ssl.endpoint.identification.algorithm", sourceConfig.getSslEndpointIdentificationAlgorithm());
```
[the remark this answers is not in the document]
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java]: we removed it.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java [hunk 7]
```diff
@@ -51,6 +51,26 @@ public class KafkaSourceConfig {
     @JsonProperty("acknowledgments_timeout")
     private Duration acknowledgementsTimeout = DEFAULT_ACKNOWLEDGEMENTS_TIMEOUT;
 
+    @JsonProperty("serde_format")
+    private String serdeFormat;
```
[the remark this answers is not in the document]
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java]: We implemented an ENUM for plaintext and avro records.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java (hunk 1 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java:59]: We made it as an ENUM.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java (hunk 2 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java]: This configuration "ssl_endpoint_identification_algorithm" is optional and removed from the config file

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java (hunk 4 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java:537]: This piece of code is required for the current implementation because we use InetAddress.getByName() for checking the server connectivity and  InetAddress.getByName() requires single server IP details at a time.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java (hunk 3 above)
Comment (ajeeshakd) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/source/KafkaSource.java:304]: Working on this. We create dedicated threads for checking each Kafka Server availability and their connection retry part.

**Review by kkondaka — APPROVED** [review]

**Review by hshardeesi — APPROVED** [review]
It looks like if there is an exception in KafkaConsumer thread, it exits w/o logging exception. Please repro it and address it.
